server/src/main/java/org/elasticsearch/snapshots/SnapshotsService.java

```diff
@@ -1131,72 +1430,232 @@ public static boolean useIndexGenerations(Version repositoryMetaVersion) {
+                        }
+                        return ClusterState.builder(currentState).putCustom(SnapshotDeletionsInProgress.TYPE, deletions)
+                                .putCustom(SnapshotsInProgress.TYPE, updatedSnapshotsInProgress).build();
                     }
                 }
                 return currentState;
             }
 
             @Override
             public void onFailure(String source, Exception e) {
-                logger.warn(() -> new ParameterizedMessage("{} failed to remove snapshot deletion metadata", snapshotIds), e);
-                if (listener != null) {
-                    listener.onFailure(e);
+                logger.warn(() -> new ParameterizedMessage("{} failed to remove snapshot deletion metadata", deleteEntry), e);
+                runningDeletions.remove(deleteEntry.uuid());
+                final String repoName = deleteEntry.repository();
+                synchronized (currentlyFinalizing) {
+                    if (ExceptionsHelper.unwrap(e, NotMasterException.class, FailedToCommitClusterStateException.class) != null) {
+                        // Failure due to not being master any more so we don't try to do run more cluster state updates. The next master
+                        // will try handling the missing operations. All we can do is fail all the listeners on this master node so that
+                        // transport requests return and we don't leak listeners
+                        final Exception wrapped =
+                                new RepositoryException(repoName, "Failed to update cluster state during snapshot delete", e);
+                        final Deque<SnapshotFinalization> outstandingSnapshotsForRepo = snapshotsToFinalize.remove(repoName);
+                        if (outstandingSnapshotsForRepo != null) {
+                            SnapshotFinalization finalization;
+                            while ((finalization = outstandingSnapshotsForRepo.poll()) != null) {
+                                failSnapshotCompletionListeners(finalization.entry.snapshot(), wrapped);
+                            }
+                        }
+                        for (Iterator<List<ActionListener<RepositoryData>>> iterator = snapshotDeletionListeners.values().iterator();
+                             iterator.hasNext(); ) {
+                            List<ActionListener<RepositoryData>> listeners = iterator.next();
+                            iterator.remove();
+                            failListenersIgnoringException(listeners, wrapped);
+                        }
+                        assert snapshotDeletionListeners.isEmpty() :
+                                "No new listeners should have been added but saw " + snapshotDeletionListeners;
+                    } else {
+                        assert false : "Removing snapshot entry should only ever fail because we failed to publish new state";
```

Comment:
should we still call the listeners here?

Reply:
I don't know. It's a broken situation either way if we get here and come to think of it we have the same issue in the snapshot path. In the end, only a master fail-over can fix things here and it should be impossible to arrive here unless there is a bug.
If we call the listeners we'll return from the request but the cluster state is still stuck. If we don't call the listeners we won't return from the request so at least the issue is a little more visible? Seems to me it makes no difference what we do here practically.